cloudscale/resource_cloudscale_volume_test.go

```diff
@@ -0,0 +1,270 @@
+		return err
+	}
+
+	foundError := error(nil)
+	for _, s := range volumes {
+		if strings.HasPrefix(s.Name, "terraform-") {
+			log.Printf("Destroying Volume %s", s.Name)
+
+			if err := client.Volumes.Delete(context.Background(), s.UUID); err != nil {
+				foundError = err
+			}
+		}
+	}
+	return foundError
+}
+
+func TestAccCloudscaleVolume_Change(t *testing.T) {
+	var volume cloudscale.Volume
+
+	rInt := acctest.RandInt()
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckCloudscaleVolumeDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: volumeConfig_detached(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudscaleVolumeExists("cloudscale_volume.basic", &volume),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "name", fmt.Sprintf("terraform-%d", rInt)),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "type", "ssd"),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "size_gb", "50"),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "server_uuids.#", "0"),
+				),
+			},
+			{
+				Config: volumeConfig_multiple_changes(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudscaleVolumeExists("cloudscale_volume.basic", &volume),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "name", fmt.Sprintf("terraform-%d-renamed", rInt)),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "type", "ssd"),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "size_gb", "100"),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "server_uuids.#", "0"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccCloudscaleVolume_Detach(t *testing.T) {
+	var server cloudscale.Server
+	var volume cloudscale.Volume
+
+	rInt1 := acctest.RandInt()
+	rInt2 := acctest.RandInt()
+
+	serverConfig := testAccCheckCloudscaleServerConfig_basic(rInt1)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckCloudscaleVolumeDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: serverConfig + "\n" + volumeConfig_attached(rInt2),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudscaleServerExists("cloudscale_server.basic", &server),
+					testAccCheckCloudscaleVolumeExists("cloudscale_volume.basic", &volume),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "server_uuids.#", "1"),
+					assertVolumeAttached(&server, &volume),
+				),
+			},
+			{
+				Config: serverConfig + "\n" + volumeConfig_detached(rInt2),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudscaleVolumeExists("cloudscale_volume.basic", &volume),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "server_uuids.#", "0"),
```

Comment:
Not sure if it's necessary, but you could add an additional check against the cloudscale.ch API and check if the volume is indeed detached.

Reply:
I added additional checks against the cloudscale API in all places where volume existence is checked.